osc.lua: use minus character with time remaining

rhendric opened the pull request:
This PR replaces the hyphen used to indicate time remaining in the on-screen controls with a U+2212 Minus Sign. The width of the time remaining/media duration region is also increased to accommodate the additional width of the minus sign relative to the hyphen.

Tested and screenshotted on Linux.

Before (hyphen):
![before](https://user-images.githubusercontent.com/1570964/177263622-a22e37b1-12f1-4696-b6c1-d18b59b8cb77.jpg)

After (minus sign):
![after](https://user-images.githubusercontent.com/1570964/177263640-aec44648-b4f0-4693-a468-cc1d3050b794.jpg)

CogentRedTester commented:
Would the use of unicode result in the character being missing on some systems/fonts?

avih commented:
You didn't explain why...

I'm generally against it. I don't think it adds value in terms of readability, and it makes the seekbar smaller, which hurts usability IMO.

rhendric commented:
> Would the use of unicode result in the character being missing on some systems/fonts?

The OSC already uses a U+2013 En Dash to represent unavailable audio or, as seen in the screenshots, subtitle content. In practice, I would be very surprised if a system can render that character but not U+2212.

> You didn't explain why...

Sorry, I figured it was self-explanatory.

The minus sign is the correct character to use to represent subtraction or a negative number. See, for example, https://en.wiktionary.org/wiki/minus_sign:

> Often in typography (and always in computer programming), the [hyphen](https://en.wiktionary.org/wiki/hyphen) (-) is used instead of the minus sign (−); in many fonts, the symbols are barely distinguishable or not at all (although the difference is clear in Times Roman). Careful typographers use the minus sign instead of the hyphen.

I suppose I'm a ‘careful typographer’; I notice when a hyphen is used where a minus sign would be more appropriate and I find it distracting.

Other internet pedants agree!
* https://brushingupscience.com/2015/12/28/a-hyphen-is-not-a-minus-sign-and-other-often-mistaken-characters/
* https://people.physics.illinois.edu/celia/Dashes.pdf#page=7
* https://www.punctuationmatters.com/the-difference-between-a-dash-and-a-minus-sign/
* https://jakubmarian.com/hyphen-minus-en-dash-and-em-dash-difference-and-usage-in-english/

> it makes the seekbar smaller, which hurts usability

I'm very open to adjusting this, but the seekbar is such an arbitrary size that I find it difficult to believe this is a pressing concern. The entire OSC scales with the window instead of enforcing any minimum or maximum size for clickable elements. That leads me to believe that the size of these elements wasn't chosen with any particular care, so shaving off a few pixels is just as likely to be a (tiny!) move toward some global usability optimum as away from it.

rhendric commented:
> > The OSC already uses a U+2013 En Dash to represent unavailable audio or, as seen in the screenshots, subtitle content
> 
> Not sure about the unavailable audio, but the content part (I assume you refer to "ACT I - The Prophecy")

No, I'm talking about the grayed-out –/0 text, indicating that no subtitle content is available. For pure videos with no audio tracks, the same thing happens to the audio track selector. Both of those dashes come from osc.lua, specifically lines 1919 and 1938, and they are Unicode characters.

Will respond to remaining arguments later, just wanted to correct that miscommunication first.

avih commented:
Also, if you do choose to go the option way, do ensure it's handled in all layouts and not only at the bottom/top bar (I believe you're missing box and slimbox, but not sure if it's needed there).

rhendric commented:
So, first of all:

> Alternatively, if you want to make it an optional thing (which still defaults to dash), and want to update the patch accordingly, including the docs, we can do that too.

This is a generous compromise and, if none of the rest of this is convincing or you simply run out of time or patience for the debate, I will gratefully fall back on it. I say ‘generous’ because, of course, this represents a higher maintenance burden on the mpv team than simply choosing one or the other behavior. So I still believe that the best option *for you* is to merge a version of this patch without the option, and I'll continue to try to reach common understanding on this point—but again, whenever the debate no longer seems worth having to you just say so and I'll take this path.

---

> Personally I don't know how likely it is for a font to be missing the unary minus codepoint glyph, but I'm willing to bet there are more fonts missing it than fonts which miss the dash codepoint (that's not an unsolvable problem, just a factor to consider).

Assuming that by dash codepoint you mean the U+2013 En Dash (which, to reiterate, is already being used [here](https://github.com/mpv-player/mpv/blob/659ab2f946be1cf09c01d510cb3ffb59972556f6/player/lua/osc.lua#L1919) and [here](https://github.com/mpv-player/mpv/blob/659ab2f946be1cf09c01d510cb3ffb59972556f6/player/lua/osc.lua#L1938)), then I'd like to take you up on that bet. On a Linux computer, it's fairly easy to look for installed fonts that have only one of the two glyphs; the following Bash command will do it:
```bash
comm -13 <(fc-list :charset=2212 fullname) <(fc-list :charset=2013 fullname)
```

On my computer, the above command prints nothing—out of the 195 font faces I have installed, 132 of them have both U+2212 and U+2013, and the remainder have neither. This is what I would expect; both codepoints were introduced in Unicode 1.1, back in 1993. They aren't exotic. If you are also running Linux, I would be curious to know if your system yields different results. Of course this isn't a non-existence proof; I don't presume to assert that no font exists with U+2013 but not U+2212. But the probability that such a font is in use seems demonstrably lower than the probability that a font without even U+2013 is in use, and mpv already accepts that chance of a missing glyph.

> I'm not saying your opinion and those links carry no weight, just that to me it's not a clear cut sort of official "using dash is completely wrong for numbers, unary minus must always be used where possible", etc. (I did follow your links, all of them seem to be unofficial opinions).

To be fair to your side of the argument, I think the consensus among people who, as you imply, support something other than the least-effort status quo, is that using hyphen is acceptable when the minus sign isn't practical (such as when accepting typed input, as Latin-alphabet keyboard layouts historically have a key for hyphen but not for the minus sign), but that the minus sign is preferred for presentation. The Unicode name for the ASCII hyphen is even [U+002D Hyphen-Minus](https://unicodeplus.com/U+002D).

That said, I'm not sure what sort of official source you'd hope to see. I have a copy of the 15th edition of the Chicago Manual of Style in front of me, and on page 263 of this printing, it says, ‘The en dash is sometimes used as a minus sign, but minus signs and en dashes are distinct characters.’ We're talking about a hyphen glyph being used as a minus sign, which is a specific case this edition of CMoS doesn't seem to cover, but page 527 in the section on mathematical expressions displays as an example of the minus sign a glyph that is obviously distinct from the hyphen glyph in weight, dimension, and position. Unfortunately, I can't link to any of this.

But really... the Unicode standard calls U+2212 ‘[Minus Sign](https://unicodeplus.com/U+2212)’. It's the codepoint associated with the named character reference `&minus;` in the [HTML specification](https://html.spec.whatwg.org/multipage/named-characters.html#named-character-references). I'm not sure how many more ways there are to demonstrate that it is the preferred codepoint for font authors and font users to use for a glyph representing the minus sign. Do you need a citation for the fact that we call the symbol that appears on the left of negative values the ‘minus sign’? If so, are you not a native English speaker by any chance? Absolutely no disrespect meant, but that is something that I would expect every English-speaking school to teach in math classes and that I would struggle to find an ‘official’ source for—maybe I could dig up a primary-level math textbook if you want?

> Also, this is by far not the only place in mpv where a dash is used as unary minus. I don't think it would be simple to convert all and only unary minus dash to the appropriate codepoint. Most probably because C and lua negative numbers are converted to string by the language itself using a dash prefix. You'd have to agree that consistency carries a weight too.

I agree! I had an earlier version of this commit in which I changed a lot more hyphens into minus signs! I thought a more targeted patch affecting the most visible change would be easier for you to accept, but if this becomes the last objection standing I would be very happy to go correct more hyphens.

The one caveat here is that some numbers end up in non-presentational places—filenames, for example—and it would be much more obnoxious to introduce non-ASCII characters there. Messages dumped into log files might also belong in this category. I intend to account for this sort of thing if given free reign.

> It's true that we don't define a width X below which we consider it unusable. It's a progressive thing. As you can tell at your own screenshot, it's small. If you add meaningful data over it (chapter markers, demuxer cache indication, etc), then the smaller it is, the less usable it becomes, so where possible and reasonable, the preference is to not make it further smaller.

Isn't the most usability-relevant dimension of the seekbar the width? If you inspect the screenshots carefully you'll see that while the height of the seekbar does drop a little, the width actually *increases*, because decreasing the height of the entire bottom bar shrinks the other elements in the bar horizontally by more pixels than I've added to the time remaining label. So I don't think this is as cut-and-dried as you're making it out to be—some usability is lost by making the bar a little shorter, some usability is gained by making it a little wider, and arguably the latter is more important than the former up to a point.

In these screenshots, the ‘before’ bar is approximately (aliasing and JPEG artifacts make it hard to get precise measurements) 159×22 pixels, or 3498 total pixels in area, and the ‘after’ bar is approximately 167×21 pixels, or 3507 total pixels in area. It's pretty close even if you're just looking at total pixels regardless of shape (I won't claim a win for the ‘after’ bar because of margin of error). So again: if we aren't aiming for some known ideal dimensional target, it is not at all clear that this change moves us closer to or further away from that ideal.

> I believe you're missing box and slimbox, but not sure if it's needed there

All of the documented layouts were covered with this change, but thanks for the reminder and I will continue to test them as the patch is developed.

avih commented:
> this represents a higher maintenance burden

No. Maintenance burden is when the defaults change and it becomes a never ending bikeshed. Supporting your preference while keeping the current default is less time consuming in the long term (also because making it an option is trivial, code wise).

In this case I guess many won't notice, but it's my job to speak for them still, because I think for most people using these extra pixels for the seekbar is more valuable than using them for the unicode minus sign.

> Assuming that by dash codepoint you mean the U+2013

I don't. All my references to "dash" meant what's used currently, which is U+002D (also evident from the following mentions which you even quoted). As in, replacing it with what you want will make it more likely to be a missing glyph. Apologies for the inaccuracy.

> the minus sign is preferred for presentation. The Unicode name for the ASCII hyphen is even [U+002D Hyphen-Minus](https://unicodeplus.com/U+002D).

Right. I'm wlling to accept that. Preferred, but still an opinion, and the currently used 0x2D is also acceptable.

> I'm not sure what sort of official source you'd hope to see

Preferably something by Unicode, but we already know that presumably Unicode does not prohibit the use of 0x2D as minus because it's part of the codepoint name...

> I have a copy of the 15th edition of the Chicago Manual of Style in front of me, and on page 263 of this printing ...

> It's the codepoint associated with the named character reference `&minus;` in the [HTML specification](https://html.spec.whatwg.org/multipage/named-characters.html#named-character-references)

These do carry more weight for me than what I've read at the links you posted earlier.

Out of curiosity, any idea which is the most commonly used codepoint at the web to represent a unary minus?

> I would be very happy to go correct more hyphens.

Please don't. It's not worth spending the (ours and your) time and touching so much code for something which is not even broken, but merely a publication style preference.

> Isn't the most usability-relevant dimension of the seekbar the width?

It is.

> If you inspect the screenshots carefully you'll see that while the height of the seekbar does drop a little, the width actually _increases_

I did inspect it closely before my previous post, and did notice the increased width of the seekbar at the screenshot, but I considered it some inconsistency on your part when taking the screenshot, because clearly your new minus sign is wider than the current hyphen-minus sign, and I presumed the code increases the width of this field at the expense of other things, probably specifically the seekbar. I did not evaluate which dimensions get changed in practice.

> because decreasing the height of the entire bottom bar shrinks the other elements in the bar horizontally

RIght, I especially did not consider that your change actually shrinks the height.

So why _do_ you shrink the height (and all the text at the bottom line with it)? this is really undesirable... and also not explained as far as I can tell, neither at the commit message and not even at your reply above... (note to yourself, such change does need an explanation at the commit message).

Is it to make the text smaller so that more chars could fit at the line, because the bigger width of the new codepoint makes it less spacious?

> > I believe you're missing box and slimbox, but not sure if it's needed there
> 
> All of the documented layouts were covered with this change

The box and slimbox layouts are docuented before any other layouts https://mpv.io/manual/master/#on-screen-controller-layout and definitely have the "time remaining" field. The function which your patch touches is `bar_layout`, which is only used by the `topbar` and `bottombar` layouts as far as I can tell.

So I'm guessing box and slimbox still use 0x2D even with your patch.

Anyway, while sort of interesting, I prefer to not spend much more time on this. I think we both understand what the other prefer and why.

Feel free to reply to my questions, but generally speaking, If you insist on going on with this (rather than just fix your own private copy of osc.lua), do the following:
- Make it a documented option, while trying to keep code changes to minimum.
- When the option is active, make sure all layouts get the new codepoint.
- When the option is active, make sure all four layouts are still visually reasonable, sizes adjusted as needed.
- If possible, I'd prefer to make this field a bit wider at the expense of only the seekbar than to shrink the height of the whole line. But even better, it would be best if you can keep all sized unmodified somehow.

Personally I'd still prefer to not do anything at all about it, because I'm pretty sure it can't be considered broken and I prefer the consistency and bigger text/seekbar, but since you seem to feel very strongly about it, and you're not entirely wrong either, I'm willing to accept the above if you insist.

rhendric commented:
> Out of curiosity, any idea which is the most commonly used codepoint at the web to represent a unary minus?

On the web in general? Hyphen-minus, I'm sure. That's probably part of why it looks more normal to you than the wider minus sign. Wikipedia is the one notable place where I've seen the correct minus sign used somewhat consistently—see, e.g., https://en.wikipedia.org/wiki/Negative_number, or for a non-mathematical article, the temperatures in https://en.wikipedia.org/wiki/Saskatchewan.

The mathematical typesetting programs TeX and LaTeX, and their web-based front-ends and derivatives, do the right thing too. But I don't know if that's the sort of thing you were asking about.

(By the way, just to be explicit: the *reason* that U+2212 should be preferred to U+002D is that font authors know to give − the same width, weight, and vertical position as the horizontal bar of +. This makes mathematical expressions (*x* + *y* − *z*) look nicer, and makes tables of figures like

| Code | Growth |
|----|----|
| ABC | +0.3% |
| DEF | +1.7% |
| GHI | −2.0% |

line up in a maximally-readable way. U+002D is generally rendered as a typographical hyphen by fonts rather than a typographical minus sign because hyphens are more common, so even though the Unicode name reflects its *semantic* flexibility when interpreting an electronic document that uses it, from a typographical or graphic design perspective, U+002D is unambiguously not the ideal glyph to use for this purpose. Admittedly, mpv is not rendering columns of numbers or mathematical expressions, so those arguments don't directly apply—I'm just trying to justify my use of language like ‘correct’ and ‘the right thing’, even in the context of what the Unicode standard says.)

> So why _do_ you shrink the height (and all the text at the bottom line with it)? this is really undesirable... and also not explained as far as I can tell, neither at the commit message and not even at your reply above... (note to yourself, such change does need an explanation at the commit message).

Oh, I didn't mean to! This looks like it's an indirect consequence of pre-existing code that scales the entire UI down if the window is smaller than the bar's preferred minimum width. I'm not entirely sure how it works, to be honest; it seems to behave differently with input files of different natural resolutions or aspect ratios or something, and I haven't fully deciphered what variables in the drawing code correspond to what things. You can see from the patch, though, that I haven't added any code for shrinking height; it just falls out of the existing layout code, whether that's right or wrong.

And on that topic...

>   * If possible, I'd prefer to make this field a bit wider at the expense of only the seekbar than to shrink the height of the whole line. But even better, it would be best if you can keep all sized unmodified somehow.

There are definitely smarter things I could imagine that layout code doing. For starters, it could call out to a system library to measure the actual required width for the textual elements with the given font rather than guessing. Different values of `--osd-font` definitely make that bar more or less cramped-looking right now. It could also adapt to the presence or absence of the minus sign when toggling between time-remaining and total-duration modes. Either or both of those would make more sense to me to do than spending more time thinking about exactly how many pixels to move around based on guesses and compromises, if we are really concerned about the best allocation of the bar's real estate. Let me know what you think is best though.

avih commented:
This is becoming worse... and I'm nearer to just say "no" and close this PR.

- This should be an osc option, not global osd option, because I already made it clear earlier that we're not changing it globally for every OSD output of mpv.
- There's no need for interface changes line.
- The code changes became bigger and more intrusive. I need the patch as small as possible. It's not your fault that the OSC code is already very delicate, but as-small-as-possible changes is really preferred.
- I still don't see an explanation for the size changes. Specifically, what gets broken without the size changes, and how the size changes unbreak it? This needs explanation because I explicitly requested it earlier, and also because I explicitly preferred no size changes, and the fact that you still kept them suggests you consider them very important, but still without explanation.

As far as I can tell the size changes don't actually fix anything. Specifically it doesn't fix the a-symmetry in spacing (because the different widths are not taken into account to begin with, regardless of which minus glyph is used). I also think the right-tc (negative) already exceeds the specified size, considering the spacing to the right is smaller than all other spaces.

I.e. without further changes (which your patch rightly _doesn't_ make), for the same timestamp value (except sign) the spacing after the righ-tc will always be smaller than the space before the left-tc, by the width of whatever minus glyph is used. Reserving more space for the tc fields simply makes the visible spaces bigger, but still equally a-symetric (in absolute pixels). I also think that the right-tc already exceeds the designated size anyway, even befroe your patch.

So how about simply just changing the glyph (when the option is active) and that's it? The a-symetry remains the same anyway (by the width of the minus glyph, which is more visible with wider glyph).

Please don't post a new patch before we agree on what should change.

avih commented:
> Sorry, honest mistake

No worries. The option should be controlled via `osc.lua` itself, as a script-opt value (e.g. maybe `typminus` or some such at `user_opts` at `osc.lua`).

> Without additional space, the time-remaining indicator appears to run into the audio track selector.

Right, the spacing is indeed very tight in your screenshot, but then again, in your first post, the image with the original minus of 0x2D already has much less spacing than desired - compare the space after it to the spaces between "chunks" elsewhere at the bottom line. Notice that the 2nd image with the "solution" is not much better in this regard either. That's why I said "I also think the right-tc (negative) already exceeds the specified size, considering the spacing to the right is smaller than all other spaces."

I think there are two issues here (even before your patch):
1. The tcw size (like all other element sizes) ignores actual font metrics, and hopes the font fits nicely enough. Your font seem wider than mine, so for you the lack of spacing is more visible than for others.
2. The tcw size doesn't take into account the fact that the right-tc is wider due to an additional minus glyph. If the spacing is plenty then it's not very noticeable, but the fatter the font is, and the wider the minus glyph is, the asymetry (and incorrect spacing) becomes more noticeable.

Your original solution of simply increasing tcw by some arbitrary value does not fix the problems above, but it does increase the spacing (and shrinks the font size), which does make the issues less noticeable.

Overall, considering the various issues, and solution approaches, I think I prefer your original solution (when active, change the glyph and add some arbitrary value only to tcw) over your current patch which meddles with sizes much more than I'd like (no offence, the whole of `osc.lua` is meddling with stuff, which is why I'd like add as little meddling as possible).

Can you live with that? And post a patch which implemets it at your next comment? (don't force push yet)

rhendric commented:
> Can you live with that?

*I* can, but I changed my approach in response to your feedback. Changing `tcW` also changes `minW`, the minimum width of a video's resolution before the UI starts getting scaled down. `minW` contains a term of `tcW` multiplied by 4—the left timestamp, the right timestamp, and, I assume, the remaining two `tcW` lengths help determine the minimum width of the seekbar. So the previous approach adds 10 to `tcW` but adds 40 to `minW`, which results in more aggressive UI scaling on small videos. You used the words ‘really undesirable’ when I pointed out the impact on UI scaling, so I took this new approach which only selectively adds 10 to the right timestamp width and thus only adds 10 to `minW`. It's more disruptive to the code but less disruptive to the layout's dimensions. Knowing that, which do you prefer?

avih commented:
> I assume, the remaining two `tcW` lengths help determine the minimum width of the seekbar

I think so too, because it sort of need to be able to "hold" two timestamps while hovering it (one right-aligned, one left aligned, without getting cluttered much by the mouse pointer).

> You used the words ‘really undesirable’ when I pointed out the impact on UI scaling

Yup, but I also preferred without size changes at all, and I like the meddling with sizes in several places even less... and in retrospect it's both not _that_ much smaller, and also likely won't be seen anyway by anyone but yourself, because by default it's unmodified.

> It's more disruptive to the code but less disruptive to the layout's dimensions. Knowing that, which do you prefer?

less disruptive to the code is the lesser evil IMO in this case.

The layout dimensions are messed up anyway, and none of the solutions so far solves it great anyway, so just change as little as possible for now.

> > Can you live with that?
> 
> _I_ can

Right. That makes both of us.

So show me the patch (diff) at your next comment before you force push a new commit please.

rhendric commented:
Hope this is what you're asking for—I don't really understand the advantage of consuming this as a comment instead of a commit but happy to comply and sorry in advance if I have misunderstood the instruction.

```patch
diff --git a/DOCS/man/osc.rst b/DOCS/man/osc.rst
index 7097b2a7a9..81a6d00d9b 100644
--- a/DOCS/man/osc.rst
+++ b/DOCS/man/osc.rst
@@ -400,6 +400,12 @@ Configurable Options
     Use ``no`` to disable chapter display on hover. Otherwise it's a lua
     ``string.format`` template and ``%s`` is replaced with the actual name.
 
+``unicodeminus``
+    Default: no
+
+    Use a Unicode minus sign instead of an ASCII hyphen when displaying
+    negative values.
+
 
 Script Commands
 ~~~~~~~~~~~~~~~
diff --git a/player/lua/osc.lua b/player/lua/osc.lua
index bd9f36ec57..b684967b57 100644
--- a/player/lua/osc.lua
+++ b/player/lua/osc.lua
@@ -53,6 +53,7 @@ local user_opts = {
     chapters_osd = true,        -- whether to show chapters OSD on next/prev
     playlist_osd = true,        -- whether to show playlist OSD on next/prev
     chapter_fmt = "Chapter: %s", -- chapter print format for seekbar-hover. "no" to disable
+    unicodeminus = false,       -- whether to use the Unicode minus sign character
 }
 
 -- read options from config and command-line
@@ -1479,6 +1480,19 @@ function bar_layout(direction)
     local padY = 3
     local buttonW = 27
     local tcW = (state.tc_ms) and 170 or 110
+
+    if user_opts.unicodeminus then
+        -- The important and desired effect here is increasing the width of the
+        -- time remaining region to the right of the seekbar. This code does
+        -- that, but also increases the width of the current time region to the
+        -- left of the seekbar and the minimum width of the seekbar itself.
+        -- Those are unintended effects but it was concluded that all of these
+        -- numbers are a bodge anyway and this is the approach minimally
+        -- disruptive to the existing code. These spacing values should ideally
+        -- be determined by the metrics of the font in use.
+        tcW = tcW + 10
+    end
+
     local tsW = 90
     local minW = (buttonW + padX)*5 + (tcW + padX)*4 + (tsW + padX)*2
 
@@ -2057,13 +2071,18 @@ function osc_init()
     -- tc_right (total/remaining time)
     ne = new_element("tc_right", "button")
 
+    local minus = "-"
+    if user_opts.unicodeminus then
+        minus = "−"
+    end
+
     ne.visible = (mp.get_property_number("duration", 0) > 0)
     ne.content = function ()
         if (state.rightTC_trem) then
             if state.tc_ms then
-                return ("-"..mp.get_property_osd("playtime-remaining/full"))
+                return (minus..mp.get_property_osd("playtime-remaining/full"))
             else
-                return ("-"..mp.get_property_osd("playtime-remaining"))
+                return (minus..mp.get_property_osd("playtime-remaining"))
             end
         else
             if state.tc_ms then
```

avih commented:
FWIW, this is one approach to allow adjusting the reserved timecode width independently of the minus glyph (untested, sans the docs):
```diff
diff --git a/player/lua/osc.lua b/player/lua/osc.lua
index c0c045b2de..96615f9d3c 100644
--- a/player/lua/osc.lua
+++ b/player/lua/osc.lua
@@ -43,6 +43,7 @@ local user_opts = {
     tooltipborder = 1,          -- border of tooltip in bottom/topbar
     timetotal = false,          -- display total time instead of remaining time?
     timems = false,             -- display timecodes with milliseconds?
+    tcwidth = 100,              -- timecode logical width (compensate font size estimation)
     visibility = "auto",        -- only used at init to set visibility_mode(...)
     boxmaxchars = 80,           -- title crop threshold for box layout
     boxvideo = false,           -- apply osc_param.video_margins to video
@@ -1482,6 +1483,11 @@ function bar_layout(direction)
     local padY = 3
     local buttonW = 27
     local tcW = (state.tc_ms) and 170 or 110
+    if user_opts.tcwidth >= 50 and user_opts.tcwidth <= 200 then
+        -- adjust our hardcoded font size estimation
+        tcW = tcW * user_opts.tcwidth / 100
+    end
+
     local tsW = 90
     local minW = (buttonW + padX)*5 + (tcW + padX)*4 + (tsW + padX)*2
```

rhendric commented:
In the long run (and by ‘long’, I mean in the next two or three weeks plus whatever time code review requires), I'd like to solve the spacing problem at its root. Specifically, I propose:
* adding a direct dependency on harfbuzz (mpv already indirectly depends on harfbuzz through libass, so this should not affect end users)
* implementing a function that uses harfbuzz to compute the total logical advance of a sequence of characters in the font face and size used for OSD
* exposing that function to scripts, and specifically using it in osc.lua on dummy strings like "00:00:00" to determine the width of all text elements; I envision that this would entail changing a lot of layout-related code in that file, including computing a different width for the right timecode based on the metrics of whatever minus glyph is in use

I had thought to do that in a subsequent PR, but given how conservative you want to be with respect to changing osc.lua code, now I feel I should ask if you would ever approve such a plan.

If I have your approval and assuming that work is successful, then I don't think a `tcwidth`-like adjustment option would ever be needed, and I will change this PR to only change the glyph and not the layout and start working on the next one.

If I don't have your approval to do something like this, then it's hacks for everyone, hacks forever, and I will implement `tcwidth` following your direction.

rhendric commented:
> please replace `"−"` with `string.char(0xe2, 0x88, 0x92)`

Will do; I just want to make sure you're aware of the fact that the en dash that appears [here](https://github.com/mpv-player/mpv/blob/659ab2f946be1cf09c01d510cb3ffb59972556f6/player/lua/osc.lua#L1919) and [here](https://github.com/mpv-player/mpv/blob/659ab2f946be1cf09c01d510cb3ffb59972556f6/player/lua/osc.lua#L1938) is represented as `"–"` and not `string.char(0xe2, 0x80, 0x93)` and you don't mind the inconsistency.

avih commented on DOCS/man/osc.rst:
```diff
@@ -400,6 +400,12 @@ Configurable Options
+    Default: no
+
+    Use a Unicode minus sign instead of an ASCII hyphen when displaying
+    negative values.
```
> when displaying negative values.

Please change to something like "... ASCII hyphen at the time-remaining duration". I don't think there are other negative values at the OSC, and if there are then this commit does not modify them.

avih commented on DOCS/man/osc.rst:
```diff
@@ -315,6 +315,11 @@ Configurable Options
+``tcwidth``
+    Logical width of timecode elements, to compensate for wider or narrower
+    fonts. 100 represents the default width, and values smaller than 50 or
+    larger than 200 are ignored.
```
"Default: 100" should come first - like elsewhere.

The bigger issue though is that it doesn't explain what it's good for. The only reason this option exists is to adjust the spacing around the two timecode values due to wider-than-expected font, but I don't think a reasonable user could guess that from this documentation.

I think something along these lines could be better (rephrase/adjust as you see fit):

Adjust the spacing around the timecode values. The timecode width depends on the font, and with some fonts the spacing near the timecodes becomes too small. Use values above 100 to increase that spacing, or below 100 to decrease it (allowed: 50-200).

(I wrote this before you edited this documentation, but my comment still largely apply)

avih replied:
Also, no need for the co-authoring thing. Keep it as your commit (if you agree with it. if you don't then remove it from the PR and I'll push it myself as my own).

garoto commented:
This whole PR back-and-forth convo is just nuts for such a mundane change. Holy shit. Someone stop this man.

rhendric commented:
@avih No objections to those code changes (I'm a fan of declaring `UNICODE_MINUS` at the top level; I was inclined to do that myself but thought you'd consider that a bigger change than necessary). The second patch is now entirely your code and documentation, and to be honest the commit message contains a lot of copy editing errors that don't matter but that I'd rather not associate with my name, so maybe that should just be your commit.

You've deviated from what I expect from a GitHub PR submission by posting an alternate patch out of band (as opposed to any of the in-band options like adding suggestions in the code review interface or pushing your own commits), so I don't know what you expect from me at this point. Do you want me to do the mechanical work of taking that patch file and pushing it to this PR branch, or are you planning on pushing that patch outside of GitHub and closing this PR and you just wanted my eyeballs on it?

@garoto I can't tell which of the two of us participating in conversation here is the man who must be stopped; if it's me sorry for engaging to an unreasonable degree I guess? I'm getting the sense that community norms on this project are maybe not what I'm used to.

Motophan commented:
This is the update I have been waiting for. I am all for correctness, especially the technical aspect of it.

avih commented:
> No objections to those code changes (I'm a fan of declaring `UNICODE_MINUS` at the top level; I was inclined to do that myself but thought you'd consider that a bigger change than necessary).

Good, thanks.

> The second patch is now entirely your code and documentation, and to be honest the commit message contains a lot of copy editing errors that don't matter but that I'd rather not associate with my name, so maybe that should just be your commit.

I completely don't mind who's name is on that commit, but I do want you to be happy with it, again, regardless of who's name is on it.

So I'd appreciate your opinion/preference/corrections about the updated name, the modified docs, and the commit message.

> You've deviated from what I expect from a GitHub PR submission by posting an alternate patch out of band (as opposed to any of the in-band options like adding suggestions in the code review interface or pushing your own commits)

I actually do consider it in-band. It's public, and at the PR tracker, and I thought I'd be less noisy this way. Nevertheless, in the future I'd try keep it more in band with you.

> so I don't know what you expect from me at this point. Do you want me to do the mechanical work of taking that patch file and pushing it to this PR branch, or are you planning on pushing that patch outside of GitHub and closing this PR and you just wanted my eyeballs on it?

If you endorsed/approved both commits the way you did with the first, then I'd have pushed it myself to master and linked here for posterity and closed this PR, extactly to avoid ongoing ping-pong of me requesting and you updating the PR.

As it stands, you're not happy with the second commit, so I ask your help to improve it (or revert it if you prefer).

Once you're happy with it, I don't care how it gets pushed. Either I push both with you as author, or I push both with me as author of the 2nd, or you update the PR with whatever and I merge it, etc. It just doesn't matter.

Personally, once you're happy with both, my preference would be to push both with your name, because it's your PR. But if you have any other preference then I'd be fine with that too.

Motophan commented:
LGTM!!!!!

rhendric commented:
I just pushed your patch with the following changes:
* Second commit message reworded
* Small edit to first sentence in your documentation for `tcspace`:
```diff
-    Adjust spacing around timecodes (time-remaining etc) at the ``bottombar``
-    and ``topbar`` layouts. [...]
+    Adjust space reserved for timecodes (current time and time remaining) in
+    the ``bottombar`` and ``topbar`` layouts. [...]
```

avih commented:
Thanks for the patches and patience.
